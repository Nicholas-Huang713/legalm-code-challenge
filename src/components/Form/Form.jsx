import { useState, useEffect } from "react";
import { getRandomDog } from "../../constants/randomDogInfo";
import { useSelector } from "react-redux";

import "./Form.scss";

const initialFormData = {
  name: "",
  exp: 0,
};
const initialDogData = {
  img: "",
  name: "",
  food: "",
};

const Form = ({ isEditMode, sendDataToServer, handleCloseForm }) => {
  const [formData, setFormData] = useState(initialFormData);
  const [errors, setValErrors] = useState({});
  const [ownedDog, setOwnedDog] = useState(initialDogData);
  const [randomDog, setRandomDog] = useState(initialDogData);
  const [apiError, setApiError] = useState(null);
  const currentDogWithOwnerInfo = useSelector((state) => state.dog.dogs);

  const handleChange = (e) => {
    const { name, value } = e.target;
    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const validate = () => {
    const errors = {};
    if (!formData.name) {
      errors.name = "Name is required";
    }
    if (!formData.exp) {
      errors.exp = "Experience is required";
    }
    return errors;
  };

  const resetInitialInputs = () => {
    setValErrors({});
    setApiError(null);
    setFormData(initialFormData);
    setRandomDog(initialDogData);
    setOwnedDog(initialDogData);
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const validationErrors = validate();
    if (Object.keys(validationErrors).length > 0) {
      setValErrors(validationErrors);
    } else {
      const dataToSend = {
        dog: isEditMode ? ownedDog : randomDog,
        owner: formData,
      };
      console.log(dataToSend);
      try {
        await sendDataToServer(dataToSend);
        resetInitialInputs();
        handleCloseForm();
      } catch (e) {
        setApiError(e.message);
      }
    }
  };

  useEffect(() => {
    if (isEditMode && currentDogWithOwnerInfo.owners) {
      const { dog, owners } = currentDogWithOwnerInfo;
      setFormData(owners[0]);
      setOwnedDog(dog);
    } else {
      const newRandomDog = getRandomDog();
      setRandomDog(newRandomDog);
    }
    return () => resetInitialInputs();
  }, [currentDogWithOwnerInfo]);

  return (
    <form onSubmit={handleSubmit}>
      <header>
        <h2>{isEditMode ? "Edit Profile" : "Adopt a Dog"}</h2>
      </header>
      {apiError && <div className="form-error">{apiError}</div>}
      <div className="form-input-container">
        <div>
          <label htmlFor="name">Name </label>
          <input
            type="text"
            id="name"
            name="name"
            value={formData.name}
            onChange={handleChange}
            className="form-input"
          />
          <div>
            {errors.name && <span style={{ color: "red" }}>{errors.name}</span>}
          </div>
        </div>
        <div>
          <label htmlFor="exp">Yrs of Exp </label>
          <input
            type="number"
            id="exp"
            name="exp"
            value={formData.exp}
            onChange={handleChange}
          />
          {errors.exp && (
            <div>
              <span style={{ color: "red" }}>{errors.exp}</span>
            </div>
          )}
        </div>
      </div>
      <div>
        <img
          src={isEditMode ? ownedDog.img : randomDog.img}
          className="form-img"
        />
      </div>
      <div className="form-dog-info">
        <div>
          <strong>Name:</strong> {isEditMode ? ownedDog.name : randomDog.name}
        </div>
        <div>
          <strong>Favorite food:</strong>{" "}
          {isEditMode ? ownedDog.food : randomDog.food}
        </div>
      </div>
      <button type="submit">Submit</button>
    </form>
  );
};

export default Form;
