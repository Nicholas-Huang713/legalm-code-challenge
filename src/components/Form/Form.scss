.form-img {
  max-width: 100%;
  height: auto;
  display: block;
  margin: 0 auto;
  max-height: 300px; // Adjust as needed to fit within the form
  object-fit: contain;
}

.form-dog-info {
  display: flex;
  justify-content: center;
  gap: 10px;
}

.form-error {
  color: red;
}

.form-input-container {
  display: flex;
  flex-direction: column;
  align-items: center;

  div {
    padding: 3px;
  }
}
