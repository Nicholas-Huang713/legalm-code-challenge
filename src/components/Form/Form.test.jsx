// Form.test.jsx
import { render, screen, fireEvent } from "@testing-library/react";
import Form from "./Form";

const handleCloseForm = vi.fn();
const mockSendDataToServer = vi.fn();

const renderNewOwnerForm = () => {
  return render(
    <Form
      isEditMode={false}
      sendDataToServer={mockSendDataToServer}
      handleCloseForm={handleCloseForm}
    />
  );
};
const renderEditOwnerForm = () => {
  return render(
    <Form
      isEditMode={true}
      sendDataToServer={mockSendDataToServer}
      handleCloseForm={handleCloseForm}
      currentDogWithOwnerInfo={{
        owners: [{ name: "jane", exp: 5 }],
        dog: { img: "imgUrl", name: "peanut", food: "apples" },
      }}
    />
  );
};

describe("Form Component", () => {
  it("renders Form and submits data", async () => {
    const { getByLabelText, getByText } = renderNewOwnerForm();
    fireEvent.change(getByLabelText(/Name/i), {
      target: { value: "John Doe" },
    });
    fireEvent.change(getByLabelText(/Yrs of Exp/i), {
      target: { value: "5" },
    });
    fireEvent.click(getByText(/Submit/i));
    expect(mockSendDataToServer).toHaveBeenCalled();
  });

  it("error messages appear inputs left blank", async () => {
    renderNewOwnerForm();
    fireEvent.click(screen.getByText(/Submit/i));
    expect(screen.getByText("Name is required")).toBeInTheDocument();
    expect(screen.getByText("Experience is required")).toBeInTheDocument();
  });

  it("error messages appear inputs left blank", async () => {
    renderEditOwnerForm();
    const nameInput = screen.getByLabelText(/Name/i);
    const expInput = screen.getByLabelText(/Yrs of Exp/i);
    expect(nameInput.value).toBe("jane");
    expect(expInput.value).toBe("5");
  });
});
