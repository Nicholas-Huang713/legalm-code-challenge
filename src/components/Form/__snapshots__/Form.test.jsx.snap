// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Form Component > renders Form and submits data 1`] = `
<div>
  <form>
    <header>
      <h2>
        Adopt a Dog
      </h2>
    </header>
    <div
      class="form-input-container"
    >
      <div>
        <label
          for="name"
        >
          Name 
        </label>
        <input
          class="form-input"
          id="name"
          name="name"
          type="text"
          value="John Doe"
        />
        <div />
      </div>
      <div>
        <label
          for="exp"
        >
          Yrs of Exp 
        </label>
        <input
          id="exp"
          name="exp"
          type="number"
          value="5"
        />
      </div>
    </div>
    <div>
      <img
        class="form-img"
        src="https://images.dog.ceo/breeds/mexicanhairless/n02113978_1069.jpg"
      />
    </div>
    <div
      class="form-dog-info"
    >
      <div>
        <strong>
          Name:
        </strong>
         
        Charlie
      </div>
      <div>
        <strong>
          Favorite food:
        </strong>
         
        peanuts
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
